File: sdk/reproit-tui-ts/test/invariant.test.ts
// Dogfood the app-invariant oracle both directions. Under the fuzzer (the
// REPROIT_INVARIANT_FILE env var is set) a violating state appends a
// REPROIT_INVARIANT marker to that file, which the TUI backend re-emits as
// EXPLORE:INVARIANT; a clean state and a production run (no env var) append
// nothing.
//
// Runs with Node's built-in test runner, node >= 22 (TS types stripped):
//   node --test sdk/reproit-tui-ts/test/invariant.test.ts
//
// No em dashes anywhere, per project rules.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { Reporter } from "../reporter.ts";

function markers(path: string): Array<{ sig: string; items: Array<{ id: string; message: string }> }> {
  if (!existsSync(path)) return [];
  return readFileSync(path, "utf8")
    .split("\n")
    .filter((l) => l.startsWith("REPROIT_INVARIANT "))
    .map((l) => JSON.parse(l.slice("REPROIT_INVARIANT ".length)));
}

// Give the reporter's fs write (a cached dynamic import under Node ESM) a tick
// to land before asserting.
const settle = () => new Promise((r) => setTimeout(r, 20));

test("invariant reports only violations under the fuzzer", async () => {
  const path = join(tmpdir(), `reproit-inv-ts-${process.pid}-${Math.random()}.ndjson`);
  process.env.REPROIT_INVARIANT_FILE = path;
  try {
    const r = new Reporter({ appId: "t" });
    r.invariant("holds", () => true);
    r.invariant("neg", () => ({ ok: false, message: "count < 0" }));
    r.invariant("falsy", () => 0);
    r.invariant("throws", () => {
      throw new Error("kaboom");
    });
    r.observeContents("Count: -1", 0, 0, "key:Down");
    await settle();

    const m = markers(path);
    assert.equal(m.length, 1);
    assert.equal(m[0].sig, r.currentSig());
    const byId = Object.fromEntries(m[0].items.map((it) => [it.id, it.message]));
    assert.deepEqual(Object.keys(byId).sort(), ["falsy", "neg", "throws"]);
    assert.equal(byId.neg, "count < 0");
    assert.equal(byId.falsy, "");
    assert.equal(byId.throws, "kaboom");
  } finally {
    rmSync(path, { force: true });
    delete process.env.REPROIT_INVARIANT_FILE;
  }
});

test("a satisfied registry and a production run write nothing", async () => {
  const path = join(tmpdir(), `reproit-inv-ts2-${process.pid}-${Math.random()}.ndjson`);
  process.env.REPROIT_INVARIANT_FILE = path;
  try {
    const r = new Reporter({ appId: "t" });
    r.invariant("holds", () => true);
    r.observeContents("Count: 3", 0, 0, "load");
    await settle();
    assert.deepEqual(markers(path), []);
  } finally {
    rmSync(path, { force: true });
    delete process.env.REPROIT_INVARIANT_FILE;
  }

  // Inert without the gate (production): no marker even with a violation.
  const r2 = new Reporter({ appId: "t" });
  r2.invariant("violated", () => false);
  r2.observeContents("Count: 4", 0, 0, "load");
  await settle();
});

File: sdk/reproit-tui-go/invariant_test.go
package reproittui

// Dogfood the app-invariant oracle both directions: a violating state appends a
// REPROIT_INVARIANT marker to the runner-provisioned file (which the TUI backend
// re-emits as EXPLORE:INVARIANT), a clean state and a production run (no gate)
// append nothing.

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func markerItems(t *testing.T, path string) []map[string]any {
	t.Helper()
	b, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var out []map[string]any
	for _, ln := range strings.Split(string(b), "\n") {
		rest, ok := strings.CutPrefix(ln, "REPROIT_INVARIANT ")
		if !ok {
			continue
		}
		var v map[string]any
		if err := json.Unmarshal([]byte(rest), &v); err != nil {
			t.Fatalf("bad marker json: %v", err)
		}
		out = append(out, v)
	}
	return out
}

func TestInvariantReportsOnlyViolationsUnderTheFuzzer(t *testing.T) {
	path := filepath.Join(t.TempDir(), "inv.ndjson")
	t.Setenv("REPROIT_INVARIANT_FILE", path)

	r := New(Config{AppID: "demo"})
	r.Invariant("holds", func() error { return nil })
	r.Invariant("neg", func() error { return errors.New("count < 0") })
	r.Invariant("boom", func() error { panic("kaboom") })

	r.ObserveContents("Count: -1", 0, 0, "key:Down")

	markers := markerItems(t, path)
	if len(markers) != 1 {
		t.Fatalf("want one marker, got %d: %v", len(markers), markers)
	}
	if markers[0]["sig"] != r.CurrentSig() {
		t.Fatalf("marker sig %v != current sig %v", markers[0]["sig"], r.CurrentSig())
	}
	items := markers[0]["items"].([]any)
	got := map[string]string{}
	for _, it := range items {
		m := it.(map[string]any)
		got[m["id"].(string)] = m["message"].(string)
	}
	if len(got) != 2 || got["neg"] != "count < 0" || got["boom"] != "kaboom" {
		t.Fatalf("unexpected violations: %v", got)
	}
	if _, ok := got["holds"]; ok {
		t.Fatalf("a holding invariant must not be reported")
	}
}

func TestInvariantSilentWhenCleanOrUngated(t *testing.T) {
	path := filepath.Join(t.TempDir(), "inv.ndjson")
	t.Setenv("REPROIT_INVARIANT_FILE", path)

	r := New(Config{AppID: "demo"})
	r.Invariant("holds", func() error { return nil })
	r.ObserveContents("Count: 3", 0, 0, "load")
	if got := markerItems(t, path); got != nil {
		t.Fatalf("a satisfied registry must write nothing, got %v", got)
	}

	// Inert without the gate (production).
	os.Unsetenv("REPROIT_INVARIANT_FILE")
	r2 := New(Config{AppID: "demo"})
	r2.Invariant("violated", func() error { return errors.New("bad") })
	r2.ObserveContents("Count: 4", 0, 0, "load")
}
